libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/navigation/NavigationHelper.java

```diff
@@ -99,25 +82,45 @@ static String buildInstructionString(RouteProgress routeProgress, Milestone mile
   /**
    * Calculates the distance remaining in the step from the current users snapped position, to the
    * next maneuver position.
+   *
+   * If the user is more than 1km away from the route, we are returning the total step distance.
    */
-  static double stepDistanceRemaining(Point snappedPosition, int legIndex, int stepIndex,
-                                      DirectionsRoute directionsRoute, List<Point> coordinates) {
+  static double stepDistanceRemaining(Location location, int legIndex, int stepIndex,
+                                      DirectionsRoute directionsRoute, List<Point> stepPoints) {
+    // If the linestring coordinate size is less than 2,the distance remaining is zero.
+    if (stepPoints.size() < 2) {
+      return 0;
+    }
+
+    Point locationToPoint = Point.fromLngLat(location.getLongitude(), location.getLatitude());
+
+    // Uses Turf's pointOnLine, which takes a Point and a LineString to calculate the closest
+    // Point on the LineString.
+    Feature feature = TurfMisc.nearestPointOnLine(locationToPoint, stepPoints, TurfConstants.UNIT_KILOMETERS);
+
+    // Check distance to route line, if it's too high, it makes no sense to snap and we assume the step distance is the whole distance of the step
+    Number distance = feature.getNumberProperty("dist");
```

Comment:
It feels a little bit ugly that we read a property here.

On other code parts we use something like this:
```java
double distance = TurfMeasurement.distance(locationToPoint, (Point) feature.geometry());
```

But your solution is calculate the distance only once, while my variant is calculate the distance twice.

I think you can keep this, but if we need this a second time, we should wrap up this in some util function.

Reply:
Yes, I know that this is pretty weird, but this is how you are supposed to use Turf AFAIK.  Turf.js uses the [same approach](https://turfjs.org/docs/#nearestPointOnLine).